causalpy/exp_inverse_propensity_weighting.py

```diff
@@ -0,0 +1,433 @@
+        Horvitz Thompson method"""
+        X = pd.DataFrame(self.X, columns=self.labels)
+        X["ps"] = ps
+        X[self.outcome_variable] = self.y
+        t = self.t.flatten()
+        p_of_t = np.mean(t)
+        X["i_ps"] = np.where(t == 1, (p_of_t / X["ps"]), (1 - p_of_t) / (1 - X["ps"]))
+        n_ntrt = X[t == 0].shape[0]
+        n_trt = X[t == 1].shape[0]
+        outcome_trt = X[t == 1][self.outcome_variable]
+        outcome_ntrt = X[t == 0][self.outcome_variable]
+        i_propensity0 = X[t == 0]["i_ps"]
+        i_propensity1 = X[t == 1]["i_ps"]
+        weighted_outcome1 = outcome_trt * i_propensity1
+        weighted_outcome0 = outcome_ntrt * i_propensity0
+        return weighted_outcome0, weighted_outcome1, n_ntrt, n_trt
+
+    def make_raw_adjustments(self, ps):
+        """This estimator is discussed in Aronow and
+        Miller as the simplest of base form of
+        inverse propensity weighting schemes"""
+        X = pd.DataFrame(self.X, columns=self.labels)
+        X["ps"] = ps
+        X[self.outcome_variable] = self.y
+        t = self.t.flatten()
+        X["ps"] = np.where(t, X["ps"], 1 - X["ps"])
+        X["i_ps"] = 1 / X["ps"]
+        n_ntrt = n_trt = len(X)
+        outcome_trt = X[t == 1][self.outcome_variable]
+        outcome_ntrt = X[t == 0][self.outcome_variable]
+        i_propensity0 = X[t == 0]["i_ps"]
+        i_propensity1 = X[t == 1]["i_ps"]
+        weighted_outcome1 = outcome_trt * i_propensity1
+        weighted_outcome0 = outcome_ntrt * i_propensity0
+        return weighted_outcome0, weighted_outcome1, n_ntrt, n_trt
+
+    def make_overlap_adjustments(self, ps):
+        """This weighting scheme was adapted from
+        Lucy D’Agostino McGowan's blog on
+        Propensity Score Weights referenced in
+        the primary CausalPy explainer notebook"""
+        X = pd.DataFrame(self.X, columns=self.labels)
+        X["ps"] = ps
+        X[self.outcome_variable] = self.y
+        t = self.t.flatten()
+        X["i_ps"] = np.where(t, (1 - X["ps"]) * t, X["ps"] * (1 - t))
+        n_ntrt = (1 - t[t == 0]) * X[t == 0]["i_ps"]
+        n_trt = t[t == 1] * X[t == 1]["i_ps"]
+        outcome_trt = X[t == 1][self.outcome_variable]
+        outcome_ntrt = X[t == 0][self.outcome_variable]
+        i_propensity0 = X[t == 0]["i_ps"]
+        i_propensity1 = X[t == 1]["i_ps"]
+        weighted_outcome1 = t[t == 1] * outcome_trt * i_propensity1
+        weighted_outcome0 = (1 - t[t == 0]) * outcome_ntrt * i_propensity0
+        return weighted_outcome0, weighted_outcome1, n_ntrt, n_trt
+
+    def make_doubly_robust_adjustment(self, ps):
+        """The doubly robust weighting scheme is also
+        discussed in Aronow and Miller, but a bit more generally
+        than our implementation here. Here we have specified
+        the outcome model to be a simple OLS model.
+        In this way the compromise between the outcome model and
+        the propensity model is always done with OLS."""
```

Comment:
What about having docstrings as Numpy (pymc-marketing)? The user mainly wants to know what `ps` is and what the function does. The rest is a (useful!) comment which can also be added.

Reply:
I've created a separate issue for this https://github.com/pymc-labs/CausalPy/issues/388